src/merkle.rs

```diff
@@ -372,61 +439,43 @@ impl<T: Element, A: Algorithm<T>, K: Store<T>> MerkleTree<T, A, K> {
         // that we're currently processing in the partial tree.
         let mut partial_base = 0;
 
-        // 'current_height' tracks the layer count being processed,
-        // starting from the bottom and increasing toward the root.
-        let mut current_height = 0;
-
-        let mut lemma: Vec<T> = Vec::with_capacity(self.height + 1); // path + root
-        let mut path: Vec<bool> = Vec::with_capacity(self.height - 1); // path - 1
+        let mut lemma: Vec<Vec<T>> = Vec::with_capacity(self.height + 1); // path + root
+        let mut path: Vec<usize> = Vec::with_capacity(self.height - 1); // path - 1
 
-        lemma.push(self.read_at(j)?);
+        lemma.push(vec![self.read_at(j)?]);
         while base + 1 < self.len() {
-            lemma.push(if j & 1 == 0 {
-                // j is left
-                let left_index = base + j + 1;
-
-                // Check if we can read from either the base data layer, or the cached region
-                // (accessed the same way via the store interface).
-                if left_index < data_width || left_index >= cache_index_start {
-                    self.read_at(left_index)?
-                } else {
-                    // Otherwise, read from the partially built sub-tree with a properly
-                    // adjusted index.
-                    let partial_tree_index =
-                        partial_base + j + 1 - (segment_start >> current_height);
-                    partial_tree.read_at(partial_tree_index)?
+            let lemma_start = (j / branches) * branches;
```

Comment:
I think this is an index into the tree, not the lemma. Name is slightly confusing.